xarray/core/datatree.py

```diff
@@ -510,6 +556,26 @@ def is_hollow(self) -> bool:
         """True if only leaf nodes contain data."""
         return not any(node.has_data for node in self.subtree if not node.is_leaf)
 
+    def _get_inherited_coord_var(self: DataTree, key: str) -> Variable:
+        for node in self.parents:
+            if key in node._local_coord_names:
+                return node._local_variables[key]
+
+        raise Exception(
+            "should never get here - means we didn't do our alignment / construction properly"
+        )
+
+    @property
+    def _inherited_variables(self: DataTree) -> Mapping[Hashable, Variable]:
+        return {
+            k: self._get_inherited_coord_var(k) for k in self._inherited_coord_names
+        }
```

Comment:
We build these lists of inherited variables/coords/dims/indexes dynamically, instead of trying to save them as static properties.